src/Artifacts.jl

```diff
@@ -283,53 +295,91 @@ function download_artifact(
-    # everything should be cleaned up.  Luckily, that is precisely what our
-    # `create_artifact()` wrapper does, so we use that here.
-    calc_hash = try
-        create_artifact() do dir
-            download_verify_unpack(tarball_url, tarball_hash, dir, ignore_existence=true, verbose=verbose,
-                quiet_download=quiet_download, io=io)
+    # everything should be cleaned up.
+
+    # Temporary directory where we'll do our creation business
+    temp_dir = mktempdir(artifacts_dir)
+
+    try
+        changed_symlinks_output = Pair{String,String}[]
+        copy_symlinks::Bool = @something(
+            PlatformEngines.copy_symlinks(),
+            (Sys.iswindows() && !PlatformEngines.can_symlink(temp_dir)),
+        )
+        download_verify_unpack(tarball_url, tarball_hash, temp_dir;
+            ignore_existence=true, verbose,
+            quiet_download, io, copy_symlinks, changed_symlinks_output,
+        )
+        calc_hash = SHA1(GitTools.tree_hash(temp_dir))
+        # Did we get what we expected?  If not, freak out.
+        if calc_hash.bytes != tree_hash.bytes
+            msg  = "Tree Hash Mismatch!\n"
+            msg *= "  Expected git-tree-sha1:   $(bytes2hex(tree_hash.bytes))\n"
+            msg *= "  Calculated git-tree-sha1: $(bytes2hex(calc_hash.bytes))\n"
+            msg *= "When downloading $(name)\n"
+            
+            # Since some windows systems don't support symlinks, 
+            # if this caused the mismatch, just give a warning.
+            # Pkg.jl#3643
+            ignore_hash_symlink = copy_symlinks && !isempty(changed_symlinks_output)
+            # Since tree hash calculation is still broken on some systems, e.g. Pkg.jl#1860,
+            # and Pkg.jl#2317, we allow setting JULIA_PKG_IGNORE_HASHES=1 to ignore the
+            # error and move the artifact to the expected location and return true
+            ignore_hash_env = Base.get_bool_env("JULIA_PKG_IGNORE_HASHES", false)
+            if ignore_hash_env || ignore_hash_symlink
+                if ignore_hash_symlink
+                    msg *= """
+
+                    Note: Julia cannot create symlinks, which may be the reason for the hash mismatch.
+                    One solution is to activate Developer Mode in Windows, for instructions on how to do that, see:
+                        https://learn.microsoft.com/en-us/gaming/game-bar/guide/developer-mode
+                    An alternative path is to ignore the artifact hash mismatches, 
+                    however this is generally not recommended.
+
+                    Symlinks that could not be extracted:
+                    $(sprint(io->show(io, MIME"text/plain"(), changed_symlinks_output)))
+                    """
+                end
+                if ignore_hash_env
+                    msg *= "\$JULIA_PKG_IGNORE_HASHES is set to 1:\n"
+                end
+                msg *= "\nIgnoring error and moving artifact to the expected location"
```

Comment:
Maybe this message should also explain how to delete the corrupt artifacts. Currently, if a user turns on "Developer Mode" after seeing this message the corrupt artifact will still be used.